java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/IndexSorter.java

```diff
@@ -55,14 +56,27 @@ public void sort(V vector, IntVector indices, VectorValueComparator<V> comparato
 
     this.comparator = comparator;
 
-    quickSort(0, indices.getValueCount() - 1);
+    quickSort();
   }
 
-  private void quickSort(int low, int high) {
-    if (low < high) {
-      int mid = partition(low, high, indices, comparator);
-      quickSort(low, mid - 1);
-      quickSort(mid + 1, high);
+  private void quickSort() {
+    Stack<Integer> rangeStack = new Stack<>();
```

Comment:
By using a stack here you are potentially allocating a decent amount of on-heap space?  Would it be better to allocator a buffer and maintain the stack manually (maybe wrap in an inner OffHeapIntStack class)?

Reply:
Sounds good. Thanks.
I have provided a simple implementation of OffHeapIntStack. Since it is intented to be used internally (at least for now), it is package private.